File: FRI3NDS.Angular4Template.Web/Assets/Scripts/app/services/_admin/_field.service.ts
import { Injectable } from '@angular/core';
import { DataAdapter } from "services/data.adapter";
import { Observable } from "rxjs/Observable";
import { _Field, _FieldBase } from "models/domain/_Field";
import { _FieldFilter } from "models/viewModels/_FieldViewModels";
import { _FieldType } from "models/domain/_FieldType";

@Injectable()
export class _FieldService {
	constructor(private dataAdapter: DataAdapter) { }

	query(filter?: _FieldFilter): Observable<_Field[]> {
		return this.dataAdapter.get("/api/Admin/_Field", filter)
			.map(response => response.json() as _Field[])
			.catch(error => {
				this.handleError(error);
				return Observable.throw(error);
			});
	}

	count(filter: _FieldFilter): Observable<number> {
		return this.dataAdapter.get("/api/Admin/_Field/Count", filter || new _FieldFilter())
			.map(response => response.json() as number)
			.catch(error => {
				this.handleError(error);
				return Observable.throw(error);
			});
	}

	getById(id: number): Observable<_Field> {
		return this.dataAdapter.get(`/api/Admin/_Field/${id}`)
			.map(response => response.json() as _Field)
			.catch(error => {
				this.handleError(error);
				return Observable.throw(error);
			});
	}

	GetFieldTypes(): Observable<_FieldType[]> {
		return this.dataAdapter.get(`/api/Admin/_Field/GetFieldTypes`)
			.map(response => response.json() as _FieldType[])
			.catch(error => {
				this.handleError(error);
				return Observable.throw(error);
			});
	}

	save(entity: _FieldBase): Observable<_FieldBase> {
		return this.dataAdapter.post("/api/Admin/_Field", entity)
			.map(response => response.json() as _FieldBase)
			.catch(error => {
				this.handleError(error);
				return Observable.throw(error);
			});
	}

	delete(id: number): Observable<number> {
		return this.dataAdapter.delete(`/api/Admin/_Field/${id}`)
			.map(response => response.json() as number)
			.catch(error => {
				this.handleError(error);
				return Observable.throw(error);
			});
	}

	private handleError(error: any) {
	}
}

File: FRI3NDS.Angular4Template.Core/Services/Data/_Admin/_FieldService.cs
﻿using FRI3NDS.Angular4Template.Core.Interfaces.Services.Data._Admin;
using System;
using System.Collections.Generic;
using System.Text;
using FRI3NDS.Angular4Template.Core.Interfaces.Data;
using FRI3NDS.Angular4Template.Core.Models.Domain;
using FRI3NDS.Angular4Template.Util;
using FRI3NDS.Angular4Template.Core.Models.Domain._Admin;
using FRI3NDS.Angular4Template.Core.Models.Enums;
using System.Linq;

namespace FRI3NDS.Angular4Template.Core.Services.Data._Admin
{
	public class _FieldService : DataServiceBase, I_FieldService
	{
		public _FieldService(IUnitOfWorkFactory unitOfWorkFactory)
			: base(unitOfWorkFactory)
		{
		}

		/// <summary>
		/// Получить поле по ее идентификатору.
		/// </summary>
		/// <param name="id">Идентификатор поля.</param>
		/// <returns>Пользователь.</returns>
		public _Field GetById(int id)
		{
			using (var uow = this.CreateAdminUnitOfWork())
			{
				return uow._FieldRepository.GetById(id);
			}
		}

		public List<_Field> Query(
			int? id = null,
			string name = null,
			string databaseName = null,
			int? entityId = null,
			string sortField = null,
			SortDirections? sortDirection = null,
			int pageSize = 1000,
			int pageNumber = 0)
		{
			using (var uow = this.CreateAdminUnitOfWork())
			{
				return uow._FieldRepository.Query(
					id: id,
					name: name,
					databaseName: databaseName,
					entityId: entityId,
					sortField: sortField,
					sortDirection: sortDirection,
					pageSize: pageSize,
					pageNumber: pageNumber);
			}
		}

		public int Count(
			string name = null,
			string databaseName = null,
			int? entityId = null)
		{
			using (var uow = this.CreateAdminUnitOfWork())
			{
				return uow._FieldRepository.Count(
					name: name,
					databaseName: databaseName,
					entityId: entityId);
			}
		}

		/// <summary>
		/// Получить список типов полей.
		/// </summary>
		/// <returns>Типы полей.</returns>
		public List<_FieldType> QueryFieldTypes()
		{
			using (var uow = this.CreateAdminUnitOfWork())
			{
				return uow._FieldTypeRepository.Query();
			}
		}

		/// <summary>
		/// Сохранить поле.
		/// </summary>
		/// <param name="field">Поле.</param>
		/// <param name="currentUserId">Идентификатор текущего пользователя.</param>
		/// <returns>Поле.</returns>
		public _FieldBase Save(_FieldBase field, int currentUserId)
		{
			{
				Argument.Require(currentUserId != 0, "Текущий пользователь не определен.");
				Argument.Require(field != null, "Поле пустое.");
				Argument.Require(field._EntityId > 0, "Идентификатор сущности пустой.");
				using (var uow = this.CreateAdminUnitOfWork())
				{
					if (field.IsIdentity)
					{
						List<_Field> entityFields = uow._FieldRepository.Query(entityId: field._EntityId);
						_Field identity = entityFields.FirstOrDefault(f => f.IsIdentity);
						Argument.Require(identity == null || identity.Id == field.Id, 
							$"У сущности должно быть единственное поле-идентификатор. " +
							$"В сущности \"{identity._EntityName}\" поле-идентификатор \"{identity.Name}\".");
					}
					return uow._FieldRepository.Save(field);
				}
			}
		}

		/// <summary>
		/// Удалить поле.
		/// </summary>
		/// <param name="id">Идентификатор поля.</param>
		/// <returns>Идентификатор поля.</returns>
		public int Delete(int id)
		{
			using (var uow = this.CreateAdminUnitOfWork())
			{
				return uow._FieldRepository.Delete(id);
			}
		}
	}
}
